internal/testrunner/runners/system/runner.go

```diff
@@ -629,11 +630,21 @@ func (r *runner) runTest(config *testConfig, ctxt servicedeployer.ServiceContext
 		hits, err = r.getDocs(dataStream)
 
 		if config.Assert.HitCount > 0 {
-			return hits.size() >= config.Assert.HitCount, err
-		}
+			if hits.size() == 0 {
+				return false, err
+			}
+
+			ret := hits.size() == oldHits
+			if !ret {
+				oldHits = hits.size()
+				time.Sleep(4 * time.Second)
+			}
 
+			return ret, err
```

Comment:
In this case, should we wait for the hits size to be greater than `config.Assert.HitCount`? Otherwise, if the time between the first and last event takes more than 5 secons, it won't wait for the last hits.
```suggestion
			if hits.size() < config.Assert.HitCount {
				return false, err
			}		
			time.Sleep(4 * time.Second)
			return true, err
```

Reply:
In this case if the 2 docs are ingested in a span of 5 seconds in 2 iterations , and `assert.hit_count: 2` is set in the config , this code returns true in the first round after it matches `2` , but total docs ingested are 4.

Reply:
May be can be mitigated this way - WDYT?

```diff
diff --git a/internal/testrunner/runners/system/runner.go b/internal/testrunner/runners/system/runner.go
index cc47cc9..a77af87 100644
--- a/internal/testrunner/runners/system/runner.go
+++ b/internal/testrunner/runners/system/runner.go
@@ -634,6 +634,10 @@ func (r *runner) runTest(config *testConfig, ctxt servicedeployer.ServiceContext
                       if config.Assert.HitCount > 0 {
			if hits.size() == 0 {
				return false, err
			}
 
+                       if hits.size() < config.Assert.HitCount {
+                               return false, err
+                       }
+
                       ret := hits.size() == oldHits
			if !ret {
				oldHits = hits.size()
				time.Sleep(4 * time.Second)
			}

			return ret, err
```

Reply:
The problem I see is that we are not waiting for the number of events to be at least the one we expect. So it may happen that the wait finishes, the count is not the one we expect yet, because we haven't waited enough, and then the later assertion fails, when it should have passed.

This implementation is waiting to have at least one event, and that the count of events is stable during at least 5 seconds. But I think it should also wait till the count of events is at least `assert.hit_count`, or timeout if it never is.

Imagine an scenario with `assert.hit_count: 2`. One event arrives during the first second, second one arrives in ten seconds.
* One event is ingested.
* One iteration happens, oldHits is zero, current hits is one, we set oldHits to one, wait 4 seconds and retry.
* Second iteration happens, oldHits is one, current hits is still one, we return true, the wait finishes.
* Assertion fails because there is only one hit. But it shouldn't because another one is going to arrive in ten seconds, way before the timeout.

I think that this wait should never return true if `config.Assert.HitCount` is set and the current count of hints is below this expected count.

Reply:
> The problem I see is that we are not waiting for the number of events to be at least the one we expect. So it may happen that the wait finishes, the count is not the one we expect yet, because we haven't waited enough, and then the later assertion fails, when it should have passed.

Resolved this in the latest comment - https://github.com/elastic/elastic-package/pull/1333#discussion_r1250752641

Reply:
Ah yes, sorry this would work :slightly_smiling_face: I was writing my comment at the same time and didn't see it.

Reply:
Pushed the changes now :-)